### PR TITLE
Cloud Hypervisor: driver update number 2

### PR DESCRIPTION
**Pull request by ericooper**
This PR introduces unit tests for clh driver #2205 and fixes for #2271 and #2270 noticed during development of the unit tests.

### CODE REVIEW

File: virtcontainers/clh_test.go [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+	"github.com/kata-containers/runtime/virtcontainers/store"
+	"github.com/kata-containers/runtime/virtcontainers/types"
+	"github.com/stretchr/testify/assert"
+)
+
+//
+// Cli helper functions
+//
+
+func cliToString(args []string) string {
```
Comment (jodh-intel) [inline virtcontainers/clh_test.go]: You could simplify this (or remove it?) in favour of:

```go
strings.TrimSpace(strings.Join(args, " "))
```

File: virtcontainers/clh_test.go [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+	err := clh.addDevice(Endpoint(tep), netDev)
+	assert.NoError(err)
+
+	director := &CommandLineDirector{}
+
+	cli, err := director.Build(clh.cliBuilder)
+	assert.NoError(err)
+	assert.NotNil(cli)
+
+	netarg, errr := getCliOption(cli.args, "--net")
```
Comment (jodh-intel) [inline virtcontainers/clh_test.go]: Nit: should this be `err`? (same comment for other functions below).

File: virtcontainers/clh_test.go [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+		Msize9p:       defaultMsize9p,
+		VirtioFSCache: virtioFsCacheAlways,
+	}
+}
+
+//
+// --cmdline <cmdline> Kernel command line
+//
+func TestClhCliKernelParameters(t *testing.T) {
+
```
Comment (jodh-intel) [inline virtcontainers/clh_test.go]: Nit: blank line at start of lots of functions.

**Comment by ericooper** [comment]
> Could you do `rebase` instead of `merge` I see one commit is a merge - LGTM

I'm not sure what you mean with this. The merge you see is the one  I do frequently to keep my fork up-to-date with kata-runtime/master ie in opposite direction. Shouldn't be a problem or?

**Comment by jcvenegas** [comment]
> > Could you do `rebase` instead of `merge` I see one commit is a merge - LGTM
> 
> I'm not sure what you mean with this. The merge you see is the one I do frequently to keep my fork up-to-date with kata-runtime/master ie in opposite direction. Shouldn't be a problem or?

Not a problem at all,  common practice I use to not have `merge` commits after update my branch with master is rebase master - so the PR that will be merged only include the list of patches you added but not a blocker at all :) 
 ```
git pull master
git checkout my-branch
git rebase master
```

File: virtcontainers/clh_test.go (hunk 3 above)
Comment (egernst) [reply virtcontainers/clh_test.go]: I think best practice to not start with newline.

File: virtcontainers/clh.go [hunk 4]
```diff
@@ -854,6 +863,11 @@ func (clh *cloudHypervisor) LaunchClh() (string, int, error) {
 	return errStr, cmd.Process.Pid, nil
 }
 
+// MaxAcrnVCPUs returns the maximum number of vCPUs supported
```
[the remark this answers is not in the document]
Comment (ericooper) [reply virtcontainers/clh.go]: the checks didn't report it but nevertheless this is obviously a cut and paste error and hence corrected.

File: virtcontainers/clh_test.go (hunk 1 above)
Comment (ericooper) [reply virtcontainers/clh_test.go]: removed and renamed

File: virtcontainers/clh_test.go (hunk 2 above)
Comment (ericooper) [reply virtcontainers/clh_test.go]: changed.

File: virtcontainers/clh_test.go (hunk 3 above)
Comment (ericooper) [reply virtcontainers/clh_test.go]: best practice is what's good for my eyes :) 
removed

File: virtcontainers/clh.go [hunk 5]
```diff
@@ -125,20 +127,26 @@ func (clh *cloudHypervisor) createSandbox(ctx context.Context, id string, networ
...
+		}
 
-	if clh.version.Major < supportedMajorVersion && clh.version.Minor < supportedMinorVersion {
-		errorMessage := fmt.Sprintf("Unsupported version: cloud-hypervisor %d.%d not supported by this driver version (%d.%d)",
-			clh.version.Major,
-			clh.version.Minor,
-			supportedMajorVersion,
-			supportedMinorVersion)
-		return errors.New(errorMessage)
+		if clh.version.Major < supportedMajorVersion && clh.version.Minor < supportedMinorVersion {
```
Comment (tedyu) [inline virtcontainers/clh.go]: supportedMajorVersion is 0.
How can clh.version.Major be lower than 0 ?

Comment (tedyu) [reply virtcontainers/clh.go]: I think the correct condition should be:
```
	if clh.version.Major < supportedMajorVersion || (clh.version.Major == supportedMajorVersion && clh.version.Minor < supportedMinorVersion) {
```

**Pull request merged**